# Guided Project: Exploring Ebay Car Sales Data

A dataset of used cars from a classifieds ection of the Germany eBay website.
The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).

The data dictionary provided with data is as follows:
* **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
* **name** - Name of the car.
* **seller** - Whether the seller is private or a dealer.
* **offerType** - The type of listing
* **price** - The price on the ad to sell the car.
* **abtest** - Whether the listing is included in an A/B test.
* **vehicleType** - The vehicle Type.
* **yearOfRegistration** - The year in which the car was first registered.
* **gearbox** - The transmission type.
* **powerPS** - The power of the car in PS.
* **model** - The car model name.
* **kilometer** - How many kilometers the car has driven.
* **monthOfRegistration** - The month in which the car was first registered.
* **fuelType** - What type of fuel the car uses.
* **brand** - The brand of the car.
* **notRepairedDamage** - If the car has a damage which is not yet repaired.
* **dateCreated** - The date on which the eBay listing was created.
* **nrOfPictures** - The number of pictures in the ad.
* **postalCode** - The postal code for the location of the vehicle.
* **lastSeenOnline** - When the crawler saw this ad last online.


In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

encoding: UTF-8 was unable to read without an error.

Column labels use camelcase, most column types (odometer,price) are strings, and have null values.

In [4]:
autos.rename({'yearOfRegistration':'registration_year','monthOfRegistration':' registration_month','notRepairedDamage':' unrepaired_damage','dateCreated':'ad_created', 'dateCrawled': 'date_crawled', 'vehicleType':'vehicle_type', 'powerPS':'power_ps','offerType':'offer_type','fuelType':'fuel_type','nrOfPictures':'nr_of_pictures','postalCode':'postal_code','lastSeen':'last_seen'}, axis = 1, inplace = True)

In [5]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', ' registration_month', 'fuel_type', 'brand',
       ' unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Clean column names: change camelCase to snake_case

( Use **Dataframe.rename() method** to rename the column, **axis =1 parameter** specify column axis and **inplace =True** parameter is to assign the result back to dataframe.)

In [6]:
autos.describe(include ='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer   registration_month fuel_type       brand  \
count       50000         50000.000000     45518       50000   
unique         13                  NaN         7          40   
top     150,000km                  NaN    benzin  volkswagen   
freq        32424                  NaN     30107       10687   
mean          NaN             5.723360       NaN         NaN   
std           NaN             3.711984       NaN         NaN   
min           NaN             0.000000       NaN         NaN   
25%           NaN             3.000000       NaN         NaN   
50%           NaN             6.000000       NaN         NaN   
75%           NaN             9.000000       NaN         NaN   
max           NaN            12.000000       NaN         NaN   

        unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count               40171                50000         50000.0  50000.000000   
unique                  2                   76             NaN           NaN   
top                  nein  2016-04-03 00:00:00             NaN           NaN   
freq                35232                 1946             NaN           NaN   
mean                  NaN                  NaN             0.0  50813.627300   
std                   NaN                  NaN             0.0  25779.747957   
min                   NaN                  NaN             0.0   1067.000000   
25%                   NaN                  NaN             0.0  30451.000000   
50%                   NaN                  NaN             0.0  49577.000000   
75%                   NaN                  NaN             0.0  71540.000000   
max                   NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Exploring the Odometer and Price Columns
Any examples of numeric data stored as text that needs to be cleaned like, price and odometer columns.

* Remove any non_numeric characters.
* Convert the column to a numeric dtype.
* Use DataFrame.rename() to rename the column to odemeter_km.

In [7]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(float)

In [8]:
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(float)

In [9]:
autos['price'].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

In [10]:
autos['odometer'].head()

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer, dtype: float64

In [11]:
autos.rename({'odometer':'odometer_km'}, axis =1, inplace = True)

In [12]:
autos[['price','odometer_km']].describe()

price    odometer_km
count  5.000000e+04   50000.000000
mean   9.840044e+03  125732.700000
std    4.811044e+05   40042.211706
min    0.000000e+00    5000.000000
25%    1.100000e+03  125000.000000
50%    2.950000e+03  150000.000000
75%    7.200000e+03  150000.000000
max    1.000000e+08  150000.000000

In [80]:
autos['odometer_km'].value_counts().sort_index(ascending =False).head()

150000.0    30085
125000.0     4857
100000.0     2058
90000.0      1673
80000.0      1375
Name: odometer_km, dtype: int64

In [81]:
autos['odometer_km'].value_counts().sort_index(ascending =True).head()

5000.0     785
10000.0    241
20000.0    742
30000.0    760
40000.0    797
Name: odometer_km, dtype: int64

In [63]:
autos['price'].value_counts().sort_index(ascending =False).head(20)


99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

In [83]:
autos['price'].value_counts().sort_index(ascending =True).head()


1.0    150
2.0      2
3.0      1
5.0      2
8.0      1
Name: price, dtype: int64

In [84]:
autos = autos[autos['price'].between(1,350000)]
#remove outliers from the table.

### Exploring the date columns

In [85]:
#To include missing values in the distribution and to use percentages instead of counts,
#chain the Series.value_counts(normalize=True, dropna=False) method.

autos[['date_crawled','ad_created','last_seen']].describe()


date_crawled           ad_created            last_seen
count                 46681                46681                46681
unique                45124                   74                37146
top     2016-03-29 23:42:13  2016-04-03 00:00:00  2016-04-07 06:17:27
freq                      3                 1821                    8

My observations: the ad_created column's time is all the same (00:00:00).


In [86]:
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The miniumn value is 1000 and the maxiumum value is 9999.

In [94]:
autos['registration_year'].value_counts(normalize=True).head(10)



2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

In [95]:
autos['registration_year'].value_counts().sort_index(ascending =True).head(10)



1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
Name: registration_year, dtype: int64

A car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. 

The first car was made in 1885/1886, so 1900 should be the earliest valid year.


In [89]:
#remove the values outside of 1900 and 2016.
autos = autos[autos['registration_year'].between(1900,2016)]

In [96]:
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

### Exploring Price by Brand

In [97]:
top_brand = autos["brand"].value_counts(normalize=True).head(10)

In [98]:
#use Series.index attribute to access the labels
selected_brands = top_brand[top_brand>0.05].index
selected_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

Select brands that have over a 5 percent of the total values.

In [115]:
avg_brand_price ={}
for b in selected_brands:
    brands = autos[autos["brand"] == b]
    avg_price = brands["price"].mean()
    avg_brand_price[b] = int(avg_price)

avg_brand_price

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

There are 6 brands of cars that have over 5 percent of the total values.

Audi,BMW and Mercedes Benz have the highest average price.
Ford and Opel are less expensive.
Volkswagen has the most listings with an average price of $5402.

### Storing Aggregate Data in a DataFrame

Use aggregation to understand the average mileage for those cars and if there's any visible link with mean price.


In [119]:
bmp_series = pd.Series(avg_brand_price)
print(bmp_series)

audi             9336
bmw              8332
ford             3749
mercedes_benz    8628
opel             2975
volkswagen       5402
dtype: int64


In [120]:
mean_mileage = {}
for m in selected_brand:
    brands = autos[autos["brand"] == m]
    avg_price = brands["odometer_km"].mean()
    mean_mileage[m] = int(avg_price)

In [121]:
mm_series = pd.Series(mean_mileage)
print(mm_series)

audi             129157
bmw              132572
ford             124266
mercedes_benz    130788
opel             129310
volkswagen       128707
dtype: int64


In [122]:
df = pd.DataFrame(mm_series, columns=['mean_mileage'])
df['mean_price'] = bmp_series

df


mean_mileage  mean_price
audi                 129157        9336
bmw                  132572        8332
ford                 124266        3749
mercedes_benz        130788        8628
opel                 129310        2975
volkswagen           128707        5402

The average mileage of these top 6 brands are over 12,000km, so more expensive brands doesn't mean to have less on mileages. Opel has the cheapest average price with around 12,000km miles.